### DISCUSSION TITLE
How to add Microsoft/Azure Application Insights to Remix app for monitoring UI metrics

### DISCUSSION THREAD

**Discussion by kevinreber** [remix-run/remix#6716] [Q&A]
## Question
Does anyone have any experience setting up Remix with `Application Insights` to monitor user data on their UI? Our code snippet below seems to work/monitor user metrics fine, but we'd like to see other opinions/solutions. 

**NOTE:** This question is related to monitoring user data through the UI, **not** how to [monitor user data through the server](https://github.com/remix-run/remix/discussions/4499) 

## Setting up `Application Insights` for UI metrics
We have 2 packages to help monitor user metrics using Azure's `Application Insights`
-  `@microsoft/applicationinsights-react-js`
-   `@microsoft/applicationinsights-web`

From what we understand, to use `Application Insights` everything must be initialized on the client side (Example code snippet below)
```tsx
// root.tsx   
import { ApplicationInsights } from "@microsoft/applicationinsights-web";
import { ReactPlugin } from "@microsoft/applicationinsights-react-js";

const reactPlugin = new ReactPlugin();
const appInsights = new ApplicationInsights({
    config: {
      autoTrackPageVisitTime: true,
      enableAutoRouteTracking: true,
      extensions: [reactPlugin],
      instrumentationKey: <OUR_MICROSOFT_INSTRUMENTATION_KEY_GOES_HERE>,
    },
  });

appInsights.loadAppInsights();
appInsights.trackPageView();

export let loader: LoaderFunction = async ({ request }) => {
  const authUser = await authenticator.isAuthenticated(request);

  // ⬇️ 🟡 Expose `username` so we can track user activity with App Insights
  return json({ username: authUser?.username || "" });
};

export default function App() {
  const loaderData = useLoaderData() as unknown as { username: string };
  const username = loaderData.username || "";

  React.useEffect(() => {
    if (username) {
      // ⬇️ 🟡 We need this to track our user's username in App Insights
      appInsights.setAuthenticatedUserContext(username);
    }
  }, []);

  return (
    <html lang="en">
       .....
      <body>
         // ⬇️ 🟡 Initialize using App Insights Context API. This provides metric support for all our UI routes
        <AppInsightsContext.Provider value={reactPlugin}> 
           .....
        </AppInsightsContext.Provider>
        <Scripts />
      </body>
    </html>
  );
}
```

If possible we'd like to avoid wrapping `appInsights.setAuthenticatedUserContext(username)` inside of a `useEffect()`. We've tried  `appInsights.setAuthenticatedUserContext(username)` inside the page `loader()`(example below)
```tsx
export let loader: LoaderFunction = async ({ request }) => {
  const authUser = await authenticator.isAuthenticated(request);
  const username = authUser?.username || "";
  appInsights.setAuthenticatedUserContext(username);

  return json({});
};
```
but `setAuthenticatedUserContext()` doesn't work inside of the page `loader()`

## Resources
- Setting up `App Insights` in React: https://learn.microsoft.com/en-us/azure/azure-monitor/app/javascript-framework-extensions?tabs=react
- Discord Thread: https://discord.com/channels/770287896669978684/1123666324212699166

**Comment by halljus** [comment]
Hey, I know it's been a while but... any chance you figured this out @kevinreber? 🙂

**Comment by Evanion** [comment]
I would like to know too. I'm running remix in a nx repo, and running remix-express is a hazzle with overriding commands etc.
